---
sidebar_label: Exception handling
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Exception Handler

:::note[motivation]
Most who have worked with Logic Apps is familiar with the error `An action failed. No dependent actions succeeded`. This is a very generic error that offers no clear information on what went wrong.

Invictus provides an **Exception Handler** Framework component that can be used to retrieve the actual error message from an Azure Logic App and optionally translate this to a human readable text.
:::

:::info[Access Rights]
The Exception Handler component requires the role `Logic App Contributor` and additionally `Website Contributor` (for Logic App Standard) to operate correctly. These must be assigned on the Exception Handler component to the resource group where the Logic Apps are located.
:::

## Resolve Azure Logic App exceptions
The **Exception Handler** component has as single `/api/ResolveException` endpoint to request inspection of failed Azure Logic App workflow runs. Based on the type of Logic App, the request looks different.

<Tabs groupId="la-type">
<TabItem value="consumption" label="Logic App Consumption" default>

If you want to resolve the exception of a Logic App Consumption use the following body:

| JSON property       | Required | Description                                                                                |
| ------------------- | :------: | ------------------------------------------------------------------------------------------ |
| `WorkflowRunId`     | yes      | The ID of the failed Azure Logic App workflow run.                                         |
| `WorkflowName`      | yes      | The name of the Azure Logic App to inspect.                                                |
| `SubscriptionId`    | yes      | The ID of the Azure subscription where the Logic App is located.                           |
| `ResourceGroupName` | yes      | The ID of the Azure resource group within the subscription where the Logic App is located. | 

<details>
<summary>**Request example**</summary>

``` json
// POST -> /api/ResolveException
{
  "WorkflowRunId": "08584773878007282344320782622CU00",
  "WorkflowName": "cdt-dev-we-somelogicapp",
  "SubscriptionId": "fb0519a2-b69f-4cdf-97ff-19735685b5b9",
  "ResourceGroupName": "some-resource-group",
  "Type": "Consumption"
}
```
</details>
</TabItem>
<TabItem value="standard" label="Logic App Standard">
If you want to resolve the exception of a Logic App Standard workflow use the following body:


| JSON property       | Required | Description                                                                                |
| ------------------- | :------: | ------------------------------------------------------------------------------------------ |
| `WorkflowRunId`     | yes      | The ID of the failed Azure Logic App workflow run.                                         |
| `WorkflowName`      | yes      | The name of the Azure Logic App to inspect.                                                |
| `SiteName`          | yes      | The site name of the Azure Logic App to inspect.                                           |
| `SubscriptionId`    | yes      | The ID of the Azure subscription where the Logic App is located.                           |
| `ResourceGroupName` | yes      | The ID of the Azure resource group within the subscription where the Logic App is located. | 

<details>
<summary>**Request example**</summary>

``` json
// POST -> /api/ResolveException
{
  "WorkflowRunId": "08584773878007282344320782622CU00",
  "WorkflowName": "ProcessMessage",
  "SubscriptionId": "fb0519a2-b69f-4cdf-97ff-19735685b5b9",
  "ResourceGroupName": "some-resource-group",
  "Type": "Standard",
  "SiteName": "cdt-dev-we-somelogicapp"
}
```
</details>
</TabItem>
</Tabs>

<details>
<summary>**Response example**</summary>

The function will respond with a list of errors for the failed actions, for example:
``` json
// 200 OK <- /api/ResolveException
{
    "Status": "Failed",
    "Code": "InvalidURL",
    "Description": "Http request failed with status code 'HostNotFound' and status message: 'No such host is known.'.",
    "Errors": [
        {
            "Code": "InvalidURL",
            "Description": "Http request failed with status code 'HostNotFound' and status message: 'No such host is known.'."
        }
    ]
}
```
</details>

## Stored Azure Logic App exceptions translations

:bulb: The ExceptionHandler function makes use of the [RegexTranslator](https://github.com/invictus-integration/docs-ifa/blob/master/framework/components/regextranslation.md) component to translate the original error from the Logic App.

:flashlight: To show what we mean by translating the error to a human readable text let's take the following example.

We have a Logic App that used to return the generic `An action failed. No dependent actions succeeded` error, by using the Exception Handler we have now improved this to the actual error which is `Http request failed with status code 'HostNotFound' and status message: 'No such host is known.'.`.

However, we have a business user that does not really understand what this means. By using the translation functionality of the Exception Handler function we can translate this error to something like `Application is not reachable` which the business user can clearly understand.

These translations must be set up in the `RegexTranslator` table in your Invictus storage account. Translations need to be added in the table as follows:

<Tabs groupId="la-type">
<TabItem value="consumption" label="Logic App Consumption" default>

| PartitionKey              | RowKey     | MatchPattern                           | OutputPattern                                           |
| ------------------------- |------------|----------------------------------------|---------------------------------------------------------|
| `logicappname.actionname` | Any RowKey | The pattern to be matched in the error | code(YourTranslationCode)The output translation pattern |

<details>
<summary>**Translation examples**</summary>

| PartitionKey                                    | RowKey | MatchPattern                                                                    | OutputPattern                                                |
| ------------------------------------------------ |--------|---------------------------------------------------------------------------------|--------------------------------------------------------------|
| `cdt-dev-we-somelogicapp.CallHTTP`               | 1      | No such host is known                                                           | code(Availability)Application is not reachable               |
| `cdt-dev-we-somelogicapp.Create`                 | 1      | cannot be evaluated because array index '0' cannot be selected from empty array | code(Account)The accountid for the debtor could not be found |
| `cdt-dev-we-somelogicapp.ParseJSON`              | 1      | Invalid JSON schema type: int                                                   | code(Data)The field is not an integer                        |
| `cdt-dev-we-somelogicapp.ExecuteStoredProcedure` | 1      | Procedure or function 'upsertData' expects parameter '@Ordernumber'             | code(Data)Ordernumber is not supplied but is required        |
</details>

</TabItem>
<TabItem value="standard" label="Logic App Standard">

| PartitionKey                       | RowKey     | MatchPattern                           | OutputPattern                                           |
| ---------------------------------- |------------|----------------------------------------|---------------------------------------------------------|
| `sitename.workflowname.actionname` | Any RowKey | The pattern to be matched in the error | code(YourTranslationCode)The output translation pattern |

<details>
<summary>**Translation examples**</summary>

| PartitionKey                                                  | RowKey | MatchPattern                                                                    | OutputPattern                                                |
|---------------------------------------------------------------|--------|---------------------------------------------------------------------------------|--------------------------------------------------------------|
| `cdt-dev-we-somelogicapp.ProcessMessage.CallHTTP`               | 1      | No such host is known                                                           | code(Availability)Application is not reachable               |
| `cdt-dev-we-somelogicapp.ProcessMessage.Create`                 | 2      | cannot be evaluated because array index '0' cannot be selected from empty array | code(Account)The accountid for the debtor could not be found |
| `cdt-dev-we-somelogicapp.ProcessMessage.ParseJSON`              | 3      | Invalid JSON schema type: int                                                   | code(Data)The field is not an integer                        |
| `cdt-dev-we-somelogicapp.ProcessMessage.ExecuteStoredProcedure` | 4      | Procedure or function 'upsertData' expects parameter '@Ordernumber'             | code(Data)Ordernumber is not supplied but is required        |
</details>

</TabItem>
</Tabs>

By using the translation functionality we will get back the following response:

``` json
{
   "status": "Failed",
    "code": "Availability",
    "description": "Application is not reachable",
    "errors": [
        {
            "code": "Availability",
            "description": "Application is not reachable"
        }
    ]
}
```

<details>
<summary>**Ignore Azure Logic App errors**</summary>

By using the translation functionality we can also ignore errors. This can be very useful for scenario's where Logic Apps can generate an error that is not relevant to the business and the flow can be seen as successful. 

Let's take an example where we have a Logic App that picks up a file from a network location and after it has been published will delete the file. We have seen situations at customers where this delete action will fail because the file has already been removed by another instance of the Logic App, this error can be safely ignored since the message is processed anyway.

To achieve this we can configure the translation as follows:

| PartitionKey                         | RowKey | MatchPattern               | OutputPattern |
| ------------------------------------ |--------|----------------------------|---------------|
| `cdt-dev-we-somelogicapp.DeleteFile` | 1      | There is no file with name | ignore        |

This will result in the following response:

``` json
{
    "status": "Succeeded",
    "code": "NoCodeSpecified",
    "description": "",
    "errors": []
}
```
</details>

## How to implement in your Azure Logic Apps?
A good way to implement the Exception Handler is to create a common exception handler Logic App that will call the Exception Handler component and return the response. This common Logic App can then be called by all other Logic Apps, this is a nice way to centralize your exception handling functionality.

:::tip
use a webhook to call the common exception handler Logic App so that if anything fails in the common Logic App you can resubmit this and return the response asynchronously.
:::

![Logic Apps setup with Exception Handler](/images/framework/logic-app-setup-w-exception-handler.png)

<details>
<summary>**Logic Apps Exception Handling example**</summary>

![logic-app](/images/exceptionhandler-logic-app.png)

The code for the `Scope Exception Handling` is shown below, pay attention to the `Terminate` action as this uses the information from the Exception Handler component and is nicely shown in the Invictus for Azure dashboard.

![dashboard](/images/exceptionhandler-logic-app-dashboard.png)


``` json
"Scope_ExceptionHandling": {
  "actions": {
	"ResolveException": {
	  "inputs": {
		"subscribe": {
		  "body": {
			"callbackurl": "@{listCallbackUrl()}",
			"logicApp": "@{workflow().name}",
			"resourceGroup": "@{parameters('resourceGroupName')}",
			"run": "@{workflow()['run']['name']}",
			"subscription": "@{parameters('subscriptionId')}",
			"correlationId": "@trigger().clientTrackingId"
		  },
		  "method": "POST",
		  "uri": "[listCallbackUrl(resourceId(variables('exceptionHandlerResourceGroupName'), 'Microsoft.Logic/workflows/triggers', variables('exceptionHandlerLogicAppName'), 'manual'), providers('Microsoft.Logic', 'workflows').apiVersions[0]).value]"
		},
		"unsubscribe": {}
	  },
	  "runAfter": {},
	  "type": "HttpWebhook"
	},
	"Terminate": {
	  "inputs": {
		"runError": {
		  "code": "@{if(equals(body('ResolveException').Status, 'Failed'), body('ResolveException').code, null)}",
		  "message": "@{if(equals(body('ResolveException').Status, 'Failed'), body('ResolveException').description, null)}"
		},
		"runStatus": "@{body('ResolveException').Status}"
	  },
	  "runAfter": {
		"ResolveException": [
		  "Succeeded"
		]
	  },
	  "type": "Terminate"
	}
  },
  "runAfter": {
	"Scope": [
	  "Failed",
	  "Skipped",
	  "TimedOut"
	]
  },
  "type": "Scope"
}
```
</details>

## Customization

<details>
<summary>**Affected Bicep parameters**</summary>

The following Bicep parameters control the inner workings of the **Exception Handler** component. See the [release pipeline step of the deployment of the Invictus Framework](./installation/index.mdx) to learn more.

| Bicep parameter | Default | Description |
| --------------- | ------- | ----------- |
| `exceptionHandlerScaling` | `{ cpuResources: '0.5', memoryResources: '1.0Gi', scaleMaxReplicas: 1, scaleMinReplicas: 0, concurrentRequests: 10 }` | The Container App options to control scaling. See [scaling rules in Azure Container Apps](https://learn.microsoft.com/en-us/azure/container-apps/scale-app?pivots=container-apps-bicep#custom). |
| `exceptionHandlerFunctionName` | `inv-${resourcePrefix}-exceptionhandler` | The name of the Azure Container App to be created for the **Exception Handler** component. |
</details>